substrate/frame/asset-rewards/src/lib.rs

```diff
@@ -0,0 +1,821 @@
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolRewardRateModified {
+				pool_id,
+				new_reward_rate_per_block,
+			});
+
+			Ok(())
+		}
+
+		/// Modify a pool admin.
+		///
+		/// Only the pool admin may perform this operation.
+		#[pallet::call_index(5)]
+		pub fn set_pool_admin(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			new_admin: Option<T::AccountId>,
+		) -> DispatchResult {
+			let caller = T::CreatePoolOrigin::ensure_origin(origin.clone())
+				.or_else(|_| ensure_signed(origin))?;
+
+			let mut pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin.as_ref().map_or(false, |admin| admin == &caller), BadOrigin);
+
+			pool_info.admin = new_admin.clone();
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolAdminModified { pool_id, new_admin });
+
+			Ok(())
+		}
+
+		/// Set when the pool should expire.
+		///
+		/// Currently the expiry block can only be extended.
+		///
+		/// Only the pool admin may perform this operation.
+		#[pallet::call_index(6)]
+		pub fn set_pool_expiry_block(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			new_expiry: DispatchTime<BlockNumberFor<T>>,
+		) -> DispatchResult {
+			let caller = T::CreatePoolOrigin::ensure_origin(origin.clone())
+				.or_else(|_| ensure_signed(origin))?;
+
+			let new_expiry = new_expiry.evaluate(frame_system::Pallet::<T>::block_number());
+			ensure!(
+				new_expiry > frame_system::Pallet::<T>::block_number(),
+				Error::<T>::ExpiryBlockMustBeInTheFuture
+			);
+
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin.as_ref().map_or(false, |admin| admin == &caller), BadOrigin);
+			ensure!(new_expiry > pool_info.expiry_block, Error::<T>::ExpiryCut);
+
+			// Always start by updating the pool rewards.
+			let reward_per_token = Self::reward_per_token(&pool_info)?;
+			let mut pool_info = Self::update_pool_rewards(&pool_info, reward_per_token)?;
+
+			pool_info.expiry_block = new_expiry;
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolExpiryBlockModified {
+				pool_id,
+				new_expiry_block: new_expiry,
+			});
+
+			Ok(())
+		}
+
+		/// Convenience method to deposit reward tokens into a pool.
+		///
+		/// This method is not strictly necessary (tokens could be transferred directly to the
+		/// pool pot address), but is provided for convenience so manual derivation of the
+		/// account id is not required.
+		#[pallet::call_index(7)]
+		pub fn deposit_reward_tokens(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&caller,
+				&pool_info.account,
+				amount,
+				Preservation::Preserve,
+			)?;
+			Ok(())
+		}
```

Comment:
there is no way for admin to clean an expired pool, get the consideration back, and get the remaining balance in the pot back too?

Reply:
I added the `cleanup_pool` method here https://github.com/paritytech/polkadot-sdk/pull/3926/commits/0c1cff624d20b700e65d7945349489657d06f9fb, it is only executable if pool is empty, so I had to alter the `unstake` and `harvest` calls here https://github.com/paritytech/polkadot-sdk/pull/3926/commits/eef21d748bad14b3d040e02802fb2801e85c00fc, also I made admin to be non option, otherwise there is no account to send the remaining rewards https://github.com/paritytech/polkadot-sdk/pull/3926/commits/a6183a559dd11c3708d0c42958647da587686bad . `cleanup_pool` can be called only by admin now, otherwise clean up should be only possible if expiration elapsed. Without that constrain admin can clean up the pool if there is no stakers and admin made mistake. Also note, to clean up the pool all stakers should harvest and unstake the funds, these calls can be executed only be the stakers. The existence of the admin for every pool should not be an issue for a user since admin can alter the pool only by increasing the rewards or extending the expiration of the pool.

Reply:
I'm pretty sure some people will never claim their reward. (account lost, or value too small to care).

But in our usecase it is ok. Also we can always improve on this later.

Maybe there could be a destroying phase that admin can start once pool is expired.
In this phase admin can then call a new call: `distribute_rewards_and_unstake` to reward and unstake a staker.

Reply:
I feel like it would ideal with a destroying phase.

But the logic right now doesn't have any issue to me, so if you prefer I can approve now.

Reply:
I made `unstake` and `harvest_rewards` permissionless if a pool is expired with optional `staker` parameter. what you think? https://github.com/paritytech/polkadot-sdk/pull/3926/commits/561144153e4f1d7a9bca8576567a05a8833cd870